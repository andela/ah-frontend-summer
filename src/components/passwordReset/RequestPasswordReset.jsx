import React from 'react';
import PropTypes from 'prop-types';
import {
    Form, Grid, Segment, Button, Header, Divider, Message
}from 'semantic-ui-react';
import { Link } from 'react-router-dom'
import '../../assets/styles/SignupForm.scss';

const RequestPasswordReset = (props) => {

    const {
        isDisplayed,
        onInputChange,
        disabled,
        loading,
        onSubmit,
        afterSubmit,
        showError,
        errorMsg,
        message
    } = props;

    return (
        <div className="SignupForm">
            <Grid columns={2} relaxed='very' stackable>
                <Grid.Column></Grid.Column>
                <Grid.Column padded='very'>
                    <Segment padded='very'>
                        <Header as='h1' textAlign='center'>
                            Authors Haven
                        </Header>
                        <Header as='h4' color='teal' textAlign='center'>
                            A home for the creative at heart
                        </Header>
                        <br />
                        <br />
                        <div style={{display : !isDisplayed && "none"}}>
                            <Form>
                                <Header as='h5' textAlign='center'>
                                    Enter your registered email to reset your password
                                </Header>
                                { showError ? (
                                    <Message color="red">
                                        {errorMsg}
                                    </Message>
                                ): null
                                }
                                <div className="spaceComponents">
                                    <Form.Input
                                        icon='mail'
                                        iconPosition='left'
                                        placeholder='Email'
                                        required
                                        onChange={onInputChange}
                                        name="email" />
                                </div>

                                <div className="spaceComponents">
                                    <center>
                                        <Button
                                            color='teal'
                                            size="big"
                                            disabled={disabled}
                                            loading={loading}
                                            onClick={onSubmit}
                                            className="SignupBtn">
                                            Next
                                        </Button>
                                    </center>
                                </div>
                            </Form>
                            <Divider horizontal>Or</Divider>
                            <div className="spaceComponents">
                                <Header color='teal' textAlign='center' padded="very">
                                    <Link to='/'>Home</Link> | <Link to='/login'>Login</Link>
                                </Header>
                            </div>
                        </div>
                        <div style={{display : isDisplayed && "none"}}>
                            <hr/>
                            <br />
                            <br />
                            <Header as='h3' color='teal' textAlign='center'>
                                { message }
                            </Header>
                            <br />
                            <br />
                            <hr/>
                            <br />
                            <div className="spaceComponents">
                                <center>
                                    <Link to="/">
                                        <Button
                                            color='teal'
                                            size="big"
                                            loading={loading}
                                            onClick={afterSubmit}
                                            className="SignupBtn">
                                            OK
                                        </Button>
                                    </Link>
                                </center>
                            </div>
                        </div>
                    </Segment>
                </Grid.Column>
            </Grid>
        </div>
    );
};

RequestPasswordReset.propTypes = {
    isDisplayed: PropTypes.bool.isRequired,
    onInputChange: PropTypes.func.isRequired,
    disabled: PropTypes.bool.isRequired,
    loading: PropTypes.bool.isRequired,
    onSubmit: PropTypes.func.isRequired,
    afterSubmit: PropTypes.func,
    showError: PropTypes.bool.isRequired,
    errorMsg: PropTypes.string,
    message: PropTypes.string
};

RequestPasswordReset.defaultProps = {
    errorMsg: "Error",
    message: "success!"
};

export default RequestPasswordReset;
