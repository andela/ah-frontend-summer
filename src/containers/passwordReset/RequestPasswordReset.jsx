import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import RequestPasswordReset
from '../../components/passwordReset/RequestPasswordReset';
import {
    requestPasswordResetAction
} from '../../store/actions/async/passwordResetActions';
import { resetState } from '../../store/actions/sync/passwordResetActions';
import '../../assets/styles/Signup.scss';

export class PasswordRequestContainer extends Component {

    state = {
        buttonDisabled: true,
        email: "",
    }

    onInputChange = (e) => {
        const emailRegex = (
            /^([a-zA-Z0-9_\-\.]{1,})@([a-zA-Z0-9_\-\.]{1,})\.([a-zA-Z]{2,5})$/
        );
        this.setState(
            {buttonDisabled: !emailRegex.test(e.target.value),
                email : e.target.value}
        )
    }

    onSubmit = () => {
        const { requestPasswordReset } = this.props;
        requestPasswordReset(this.state.email);
    };

    render() {
        const {
            error,
            message,
            loading,
            afterSubmit,
            isDisplayed
        } = this.props;

        const {
            buttonDisabled,
        } = this.state;

        let errorMsg = null
        let showError = false
        let incomingError = error

        if (incomingError !== null) {
            errorMsg = (
                ('user' in incomingError)? incomingError.user.message :
            incomingError.errors[0])
            showError = true
        }

        return (
            <div className="Signup">
                <RequestPasswordReset
                    disabled={buttonDisabled}
                    isDisplayed={isDisplayed}
                    onSubmit={this.onSubmit}
                    onInputChange={this.onInputChange}
                    message={message}
                    loading={loading}
                    errorMsg={errorMsg}
                    showError={showError}
                    afterSubmit={afterSubmit}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        message: state.passwordReset.message,
        isDisplayed: state.passwordReset.msgDisplayed,
        loading: state.passwordReset.loading,
        error: state.passwordReset.error
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        requestPasswordReset : (email) => {
            dispatch(requestPasswordResetAction(email))
        },
        afterSubmit : () => dispatch(resetState())
    };
};

PasswordRequestContainer.propTypes = {
    isDisplayed: PropTypes.bool.isRequired,
    message: PropTypes.string,
    loading: PropTypes.bool,
    error: PropTypes.object,
    requestPasswordReset : PropTypes.func,
    afterSubmit : PropTypes.func
};

PasswordRequestContainer.defaultProps = {
    loading: false,
    message: "success",
    error:null
};

export default connect(mapStateToProps, mapDispatchToProps)(
    PasswordRequestContainer
);
