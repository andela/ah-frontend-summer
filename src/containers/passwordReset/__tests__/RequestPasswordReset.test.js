import React from "react";
import { shallow } from 'enzyme';
import { PasswordRequestContainer } from "../RequestPasswordReset";

describe('Request password reset container tests', () => {
    it('renders container consistently', () => {
        const Wrapper = shallow(<PasswordRequestContainer error={null} isDisplayed={false} />);
        expect(Wrapper).toMatchSnapshot();
    });

    it('renders container correctly with fetch error', () => {
        const Wrapper = shallow(<PasswordRequestContainer error={{errors:["error"]}} isDisplayed={true} />);
        expect(Wrapper).toMatchSnapshot();
    });
});
