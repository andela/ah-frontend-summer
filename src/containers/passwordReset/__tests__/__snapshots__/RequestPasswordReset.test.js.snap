// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request password reset container tests renders container consistently 1`] = `
<div
  className="Signup"
>
  <RequestPasswordReset
    disabled={true}
    errorMsg={null}
    isDisplayed={false}
    loading={false}
    message="success"
    onInputChange={[Function]}
    onSubmit={[Function]}
    showError={false}
  />
</div>
`;

exports[`Request password reset container tests renders container correctly with fetch error 1`] = `
<div
  className="Signup"
>
  <RequestPasswordReset
    disabled={true}
    errorMsg="error"
    isDisplayed={true}
    loading={false}
    message="success"
    onInputChange={[Function]}
    onSubmit={[Function]}
    showError={true}
  />
</div>
`;
